### PR TITLE
Refactor IMWDomainCreator

### LINKED ISSUE #20032: Code repetition in IMWDomainCreator
The method `IMWDomainCreator::getXInterval` contains a large nested 'if' statement with many repeated lines, making it difficult to understand the program flow. Repeated lines should be extracted as much as possible to make code more readable.

#### Expected behaviour

Behaviour of the program should not change, and all tests should still pass.

### PR DESCRIPTION
**Pull request by josephframsay**
#### Description of work ####
Refactored a large, slightly confusing nested if statement in getXInterval to make the function more readable.

**To test:**

Run CurveFitting unit tests - they should pass as they did before the changes were made.

Fixes #20032 . 

**Release Notes** 
*Does not need to be in the release notes, as the user should notice no changes*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by brandonhewer — APPROVED** [review]
Code still works as intended and refactoring looks good.

**Comment by martyngigg** [comment]
@mantid-roman Could you take a quick look here and check your happy with these changes? The PR diff doesn't update dynamically so [this view](https://github.com/mantidproject/mantid/compare/master...20032_refactor_IMWDomainCreator) is now a proper reflection of these changes after ~~#20008~~ was merged.

File: Framework/CurveFitting/src/IMWDomainCreator.cpp [hunk 1]
```diff
@@ -145,45 +145,42 @@ std::pair<size_t, size_t> IMWDomainCreator::getXInterval() const {
 
   setParameters();
 
-  // find the fitting interval: from -> to
+  // From points to the first occurrence of StartX in the workspace interval.
+  // End points to the last occurrence of EndX in the workspace interval.
```
Comment (mantid-roman) [inline Framework/CurveFitting/src/IMWDomainCreator.cpp]: I don't think these two lines are totally correct. "from" is an iterator pointing to the first value in the X vector that falls into the interval. "to" is an iterator to the last X value in the interval.

File: Framework/CurveFitting/src/IMWDomainCreator.cpp [hunk 2]
```diff
@@ -145,45 +145,42 @@ std::pair<size_t, size_t> IMWDomainCreator::getXInterval() const {
...
+    from = std::lower_bound(X.begin(), X.end(), m_startX);
+    to = std::upper_bound(from, X.end(), m_endX);
+  } else { // x is descending
+    if (m_startX < m_endX) {
+      std::swap(m_startX, m_endX);
     }
+    from = std::lower_bound(X.begin(), X.end(), m_startX, greaterIsLess);
+    to = std::upper_bound(from, X.end(), m_endX, greaterIsLess);
+  }
+
```
Comment (mantid-roman) [inline Framework/CurveFitting/src/IMWDomainCreator.cpp]: I think the changes make sense.

File: Framework/CurveFitting/src/IMWDomainCreator.cpp (hunk 1 above)
Comment (martyngigg) [inline Framework/CurveFitting/src/IMWDomainCreator.cpp]: @mantid-roman Would you like that commenting updating but apart from that this is okay?

Comment (mantid-roman) [inline Framework/CurveFitting/src/IMWDomainCreator.cpp]: Yes

**Comment by martyngigg** [comment]
Ah, it got merged...

**Comment by ianbush** [comment]
Ah, sorry, my bad. I saw the most recent comment of 'I think the changes make sense.' and thought it was a general comment, not realising it was to a specific bit.

**Comment by mantid-roman** [comment]
It'a all right

**Comment by martyngigg** [comment]
It's only a small comment change so no real harm done :)

**Pull request merged**